README.md
[![Build Status](https://travis-ci.com/kurtlawrence/cansi.svg?branch=master)](https://travis-ci.com/kurtlawrence/cansi)
[![Latest Version](https://img.shields.io/crates/v/cansi.svg)](https://crates.io/crates/cansi)
[![Rust Documentation](https://img.shields.io/badge/api-rustdoc-blue.svg)](https://docs.rs/cansi)
[![codecov](https://codecov.io/gh/kurtlawrence/cansi/branch/master/graph/badge.svg)](https://codecov.io/gh/kurtlawrence/cansi)

# **C**atergorise **ANSI** - ANSI escape code parser and categoriser

See the [rs docs.](https://docs.rs/cansi/)
Look at progress and contribute on [github.](https://github.com/kurtlawrence/cansi)

`cansi` will parse text with ANSI escape sequences in it and return a deconstructed 
text with metadata around the colouring and styling. `cansi` is only concerned 
with `CSI` sequences, particuarly the `SGR` parameters. `cansi` will not construct
escaped text, there are crates such as [`colored`](https://crates.io/crates/colored)
that do a great job of colouring and styling text.

# Example usage

> This example was done using the `colored` crate to help with constructing the escaped 
> text string. It will work with other tools that inject escape sequences into text strings (given they follow [ANSI specification](https://en.wikipedia.org/wiki/ANSI_escape_code)).

```rust
extern crate cansi;
extern crate colored;

use cansi::*;
use colored::Colorize;
use std::io::Write;

let v = &mut Vec::new();
write!(
  v,
  "Hello, {}{}{}{}{}{}",
  "w".white().on_red(),
  "o".cyan().on_green(),
  "r".magenta().on_yellow(),
  "l".blue().on_white(),
  "d".yellow().on_bright_cyan(),
  "!".bright_red().on_bright_yellow(),
)
.unwrap();

let text = String::from_utf8_lossy(&v);
let result = categorise_text(&text); // cansi function

assert_eq!(result.len(), 7); // there should be seven differently styled components

assert_eq!("Hello, world!", &construct_text_no_codes(&result));

// 'Hello, ' is just defaults
assert_eq!(
  result[0],
  CategorisedSlice {
    text: "Hello, ",
    start: 0,
    end: 7,
    fg_colour: Color::White,
    bg_colour: Color::Black,
    intensity: Intensity::Normal,
    italic: false,
    underline: false,
    blink: false,
    reversed: false,
    hidden: false,
    strikethrough: false
  }
);

// 'w' is coloured differently
assert_eq!(
  result[1],
  CategorisedSlice {
    text: "w",
    start: 15,
    end: 16,
    fg_colour: Color::White,
    bg_colour: Color::Red,
    intensity: Intensity::Normal,
    italic: false,
    underline: false,
    blink: false,
    reversed: false,
    hidden: false,
    strikethrough: false
  }
);
```

## Targeting no_std
This crate can use `alloc` in place of the standard library for no_std targets.
The standard library is enabled by default, so disabling default features and enabling the
`alloc` feature is required to use the crate this way.

```toml
[dependencies]
cansi = { version = "2.1.0", default-features = false, features = ["alloc"] }
```



CHANGELOG.md
# Change Log

## v2.1.0

- Added `start` and `end` byte positions to `CategorisedSlice`.

> v2.1.1
>
> - fix bug in parsing characters with more than 1 byte length.
> - general performance improvement with parsing fix.

## v2.0.0

- Moved `colored` crate out of dependencies and into dev-dependency.
- `CategorisedSlice` has `text` field which is a `&str`.
- `CategorisedSlice` internally uses `const fn`.
- Introduced my own `parse` function, which returns `Vec<Match>`
- Benchmarked and refactored functions to improve performance. [see results](https://github.com/kurtlawrence/cansi/blob/master/notes/Benchmarking.md)

## v1.1.0

- Added `line_iter()` function -- ability to iterate over `CategorisedSlices` as a line iterator.

> v1.1.1
>
> - fixed bug in iterating lines not starting a new line if followed by escape codes

## v1.0.0

- stable release


src/tests.rs
use super::*;
use colored::Colorize;

#[cfg(all(not(feature = "std"), feature = "alloc"))]
use alloc::vec::Vec;

#[test]
fn cover_other_parameters() {
    // colored doesn't always test all match arms, so i test here

    // no escape sequences
    let text = "test";
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 0,
            end: 4,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );

    // empty sequences
    let text = "\x1b[;mtest";
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 4,
            end: 8,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );

    // empty text - doesn't add it
    let text = "\x1b[;mtest\x1b[;m\x1b[;m";
    assert_eq!(categorise_text(&text[..]).len(), 1);
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 4,
            end: 8,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );

    // 22
    let text = "\x1b[1;22mtest";
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 7,
            end: 11,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );

    // 23
    let text = "\x1b[3;23mtest";
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 7,
            end: 11,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );

    // 24
    let text = "\x1b[4;24mtest";
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 7,
            end: 11,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );

    // 25
    let text = "\x1b[5;25mtest";
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 7,
            end: 11,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );

    // 27
    let text = "\x1b[7;27mtest";
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 7,
            end: 11,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );

    // 28
    let text = "\x1b[8;28mtest";
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 7,
            end: 11,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );

    // 29
    let text = "\x1b[9;29mtest";
    assert_eq!(
        categorise_text(&text[..])[0],
        CategorisedSlice {
            text: "test",
            start: 7,
            end: 11,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        }
    );
}

#[test]
fn split_on_new_line_tests() {
    fn fn_as_str(s: &str) -> (&str, Option<&str>) {
        let (first, remainder) = split_on_new_line(s);
        (&s[..first], remainder.map(|i| &s[i..]))
    }

    // no remainder
    let (first, remainder) = fn_as_str("Hello worlds");
    assert_eq!(first, "Hello worlds");
    assert_eq!(remainder, None);

    let (first, remainder) = fn_as_str("Hello worlds\n");
    assert_eq!(first, "Hello worlds");
    assert_eq!(remainder, Some(""));

    let (first, remainder) = fn_as_str("Hello worlds\r\n");
    assert_eq!(first, "Hello worlds");
    assert_eq!(remainder, Some(""));

    // some remainder
    let (first, remainder) = fn_as_str("Hello worlds\none two three");
    assert_eq!(first, "Hello worlds");
    assert_eq!(remainder, Some("one two three"));

    let (first, remainder) = fn_as_str("Hello worlds\r\none two three");
    assert_eq!(first, "Hello worlds");
    assert_eq!(remainder, Some("one two three"));

    let (first, remainder) = fn_as_str("Hello worlds\r\none\ntwo\nthree\n");
    assert_eq!(first, "Hello worlds");
    assert_eq!(remainder, Some("one\ntwo\nthree\n"));

    // no first
    let (first, remainder) = fn_as_str("\r\nHello worlds\none two three");
    assert_eq!(first, "");
    assert_eq!(remainder, Some("Hello worlds\none two three"));

    let (first, remainder) = fn_as_str("\nHello worlds\r\none two three");
    assert_eq!(first, "");
    assert_eq!(remainder, Some("Hello worlds\r\none two three"));

    let (first, remainder) = fn_as_str("\r\n");
    assert_eq!(first, "");
    assert_eq!(remainder, Some(""));
}

#[test]
fn clone_style_test() {
    use colored::*;
    let s = "hello".green();
    let c = categorise_text(&s);
    let d = c[0].clone_style("why", 0, 0);

    assert_eq!(d.text, "why");

    let e = d.clone_style("hello", 0, 5);

    assert_eq!(c[0], e);
}

#[test]
fn line_iter_test() {
    let mut green = CategorisedSlice::default_style("", 0, 0);
    let mut red = CategorisedSlice::default_style("", 0, 0);
    green.fg_colour = Color::Green;
    red.fg_colour = Color::Red;

    let cat = categorise_text("hello, world");
    let mut iter = line_iter(&cat);
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("hello, world", 0, 12)])
    );
    assert_eq!(iter.next(), None);

    let cat = categorise_text("hello, world\nhow are you");
    let mut iter = line_iter(&cat);
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("hello, world", 0, 12)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("how are you", 13, 24)])
    );
    assert_eq!(iter.next(), None);

    let s = format!("{}{}\nhow are you", "hello, ".green(), "world".red());
    let cat = categorise_text(&s);
    let mut iter = line_iter(&cat);
    assert_eq!(
        iter.next(),
        Some(vec![
            green.clone_style("hello, ", 5, 12),
            red.clone_style("world", 21, 26)
        ])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("how are you", 31, 42)])
    );
    assert_eq!(&s[5..12], "hello, ");
    assert_eq!(&s[21..26], "world");
    assert_eq!(&s[31..42], "how are you");
    assert_eq!(iter.next(), None);

    let s = format!(
        "{}{}\nhow are you\r\ntoday",
        "hello, ".green(),
        "world".red()
    );
    let cat = categorise_text(&s);
    let mut iter = line_iter(&cat);
    assert_eq!(
        iter.next(),
        Some(vec![
            green.clone_style("hello, ", 5, 12),
            red.clone_style("world", 21, 26)
        ])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("how are you", 31, 42)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("today", 44, 49)])
    );
    assert_eq!(iter.next(), None);
    assert_eq!(&s[5..12], "hello, ");
    assert_eq!(&s[21..26], "world");
    assert_eq!(&s[31..42], "how are you");
    assert_eq!(&s[44..49], "today");

    let cat = categorise_text("\n\n\n\n");
    let mut iter = line_iter(&cat);
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 0, 0)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 1, 1)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 2, 2)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 3, 3)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 4, 4)])
    );
    assert_eq!(iter.next(), None);

    let cat = categorise_text("\r\n\r\n\r\n\r\n");
    let mut iter = line_iter(&cat);
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 0, 0)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 2, 2)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 4, 4)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 6, 6)])
    );
    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("", 8, 8)])
    );
    assert_eq!(iter.next(), None);
}

#[test]
fn line_iter_newline_starts_with_esc() {
    let mut green = CategorisedSlice::default_style("", 0, 0);
    green.fg_colour = Color::Green;

    let s = format!("hello\n{}", "world".green());
    let cat = categorise_text(&s);
    let mut iter = line_iter(&cat);

    assert_eq!(
        iter.next(),
        Some(vec![CategorisedSlice::default_style("hello", 0, 5)])
    );
    assert_eq!(iter.next(), Some(vec![green.clone_style("world", 11, 16)]));
    assert_eq!(&s[11..16], "world");
}

#[test]
fn line_iter_bugs() {
    let bug_str = "\u{1b}[36mpapyrus\u{1b}[0m=> 5+6\n\u{1b}[36mpapyrus\u{1b}[0m \u{1b}[92m[out0]\u{1b}[0m: 11                                            \n\u{1b}[36mpapyrus\u{1b}[0m=>
                              \n                                                                                \n                                                                                \n
                     \n                                                                                \n                                                                                \n
            \n                                                                                \n                                                                                \n
   \n                                                                                \n                                                                                \n                                                                                \n
                                                                            \n                                                                                \n                                                                                \n
                                                                   \n                                                                                \n                                                                                \n
                                                          \n                                                                                \n                                                                                \n";

    let cat = categorise_text(bug_str);
    let mut iter = line_iter(&cat);

    let mut cyan = CategorisedSlice::default_style("", 0, 0);
    cyan.fg_colour = Color::Cyan;

    assert_eq!(
        iter.next(),
        Some(vec![
            cyan.clone_style("papyrus", 5, 12),
            CategorisedSlice::default_style("=> 5+6", 16, 22)
        ])
    );

    assert_eq!(&bug_str[5..12], "papyrus");
    assert_eq!(&bug_str[16..22], "=> 5+6");
}

#[test]
fn byte_bug() {
    let s = "ﾮ";
    let matches = parse(s);
    assert_eq!(matches, vec![]);

    let x = categorise_text(&s);
    let c = construct_text_no_codes(&x);
    assert_eq!(s, c);
}

#[test]
fn colouring_with_emojis() {
    let t = "👋, \x1b[31;4m🌍\x1b[0m!";
    let c = categorise_text(t);
    assert_eq!(construct_text_no_codes(&c), "👋, 🌍!");

    let mut c = c.into_iter();

    assert_eq!(
        c.next(),
        Some(CategorisedSlice {
            text: "👋, ",
            start: 0,
            end: 6,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        })
    );

    assert_eq!(
        c.next(),
        Some(CategorisedSlice {
            text: "🌍",
            start: 13,
            end: 17,
            fg_colour: Color::Red,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: true,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        })
    );

    assert_eq!(
        c.next(),
        Some(CategorisedSlice {
            text: "!",
            start: 21,
            end: 22,
            fg_colour: Color::White,
            bg_colour: Color::Black,
            intensity: Intensity::Normal,
            italic: false,
            underline: false,
            blink: false,
            reversed: false,
            hidden: false,
            strikethrough: false
        })
    );

    assert_eq!(c.next(), None);
}


linux-test.sh
set -e	# stop on failure
cargo +stable fmt -- --check
cargo clippy -- -D warnings
cargo test
echo "PASSED"


src/parsing.rs
#[cfg(all(not(feature = "std"), feature = "alloc"))]
use alloc::vec::Vec;

/// A match.
#[derive(Debug, PartialEq)]
pub struct Match<'t> {
    /// First byte index.
    pub start: usize,
    /// Last byte index + 1.
    pub end: usize,
    /// The text slice (ie `text[start..end]`).
    /// Note that the range is `(start..end]`.
    pub text: &'t str,
}

// ESC is 0x1b
const CSI: &str = "\x1b[";

#[inline(always)]
fn terminated_byte(byte: u8) -> bool {
    byte >= 0x40 && byte <= 0x7e
}

/// Parses ANSI escape codes from the given text, returning a vector of `Match`.
///
/// ```rust
/// let ansi_text = "Hello, \x1b[31;4mworld\x1b[0m!";
/// let parsed: Vec<_> = cansi::parse(ansi_text)
///     .into_iter()
///     .map(|m| (m.start, m.end))
///     .collect();
/// assert_eq!(
///     parsed,
///     vec![(7, 14), (19, 23)],
/// );
/// ```
pub fn parse(text: &str) -> Vec<Match> {
    let mut v = Vec::with_capacity(8);
    let csi_len = CSI.len();

    let mut s = text;
    let mut start = 0;
    let mut end = start + csi_len;

    while end <= text.len() {
        if s.starts_with(CSI) {
            // start of a CSI seq
            while end < text.len() && !terminated_byte(text.as_bytes()[end]) {
                end += 1;
            }

            let end = end + 1;

            v.push(Match {
                start: start,
                end,
                text: &text[start..end],
            });

            start = end;
        } else {
            start += s.chars().next().expect("non-empty-str").len_utf8();
        }

        s = &text[start..];
        end = start + csi_len;
    }

    v
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_test() {
        let ansi_text = "Hello, \x1b[31;4mworld\x1b[0m!";
        let parsed = parse(ansi_text);
        assert_eq!(
            parsed,
            vec![
                Match {
                    start: 7,
                    end: 14,
                    text: "\x1b[31;4m"
                },
                Match {
                    start: 19,
                    end: 23,
                    text: "\x1b[0m"
                }
            ]
        );
    }

    #[test]
    fn parse_string_with_different_chars() {
        let t = "👋, \x1b[31;4m🌍\x1b[0m!";
        let parsed = parse(t);
        assert_eq!(
            parsed,
            vec![
                Match {
                    start: 6,
                    end: 13,
                    text: "\x1b[31;4m"
                },
                Match {
                    start: 17,
                    end: 21,
                    text: "\x1b[0m"
                }
            ]
        );
    }
}


benches/benches.rs
#[macro_use]
extern crate criterion;

use colored::*;
use criterion::Criterion;

fn categorise_text_fn(c: &mut Criterion) {
    c.bench_function("fn categorise_text &str", |b| {
        let txt = "\u{1b}[91mHello, world!\u{1b}[0m";
        b.iter(|| cansi::categorise_text(&txt))
    });

    c.bench_function("fn categorise_text simple", |b| {
        let txt = format!("{}", "Hello, world!".bright_red());
        b.iter(|| cansi::categorise_text(&txt))
    });

    c.bench_function("fn categorise_text long no color", |b| {
        let s = cstr();
        let txt = cansi::construct_text_no_codes(&cansi::categorise_text(&s));
        b.iter(|| cansi::categorise_text(&txt))
    });

    c.bench_function("fn categorise_text long complex", |b| {
        let txt = cstr();
        b.iter(|| cansi::categorise_text(&txt))
    });

    c.bench_function("fn categorise_text long x 4 complex", |b| {
        let txt = format!("{}{}{}{}", cstr(), cstr(), cstr(), cstr());
        b.iter(|| cansi::categorise_text(&txt))
    });
}

fn construct_text_no_codes_fn(c: &mut Criterion) {
    c.bench_function("fn construct_text_no_codes", |b| {
        let s = cstr();
        let cat = cansi::categorise_text(&s);
        b.iter(|| cansi::construct_text_no_codes(&cat))
    });
}

fn cstr() -> String {
    format!(
        "{}{}{}{}{}{}{}{}{}{}{}",
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit.".red(),
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit.".blue(),
        "Donec vel metus nec nisl ultrices cursus.".green(),
        "In in enim eget felis elementum consectetur et nec nisi.".purple(),
        "Morbi vel sapien consectetur, tristique sem id, facilisis purus.".yellow(),
        "Vivamus bibendum nisi ac lacus euismod hendrerit vel ac lacus.".red(),
        " Nulla scelerisque ipsum eu lacus dignissim, a tempus arcu egestas.".white(),
        "Nulla scelerisque ipsum eu lacus dignissim, a tempus arcu egestas.".bright_red(),
        "Praesent lobortis quam sed erat egestas, et tincidunt erat rutrum.".bright_white(),
        "Nullam maximus mauris a ultricies blandit.".bright_green(),
        "Morbi eget neque eget neque viverra mollis in id lacus.".bright_purple(),
    )
}

criterion_group!(benches, categorise_text_fn, construct_text_no_codes_fn);
criterion_main!(benches);


src/categorise.rs
use super::*;

#[cfg(all(not(feature = "std"), feature = "alloc"))]
use alloc::vec::Vec;

const SEPARATOR: char = ';';

/// Parses the text and returns each formatted slice in order.
/// The ANSI escape codes are not included in the text slices.
///
/// Each different text slice is returned in order such that the text without the escape characters can be reconstructed.
/// There is a helper function (`construct_text_no_codes`) on `CategorisedSlices` for this.
pub fn categorise_text(text: &str) -> CategorisedSlices {
    let matches = parse(text);

    let mut sgr = SGR::default();

    let mut lo = 0;

    // will always less than or equal to matches + 1 in length, see tests
    let mut slices: Vec<CategorisedSlice> = Vec::with_capacity(matches.len() + 1);

    for m in matches {
        // add in the text before CSI with the previous SGR format
        if m.start != lo {
            slices.push(CategorisedSlice::with_sgr(
                sgr,
                &text[lo..m.start],
                lo,
                m.start,
            ));
        }

        sgr = handle_seq(&m);

        lo = m.end;
    }

    if lo != text.len() {
        slices.push(CategorisedSlice::with_sgr(
            sgr,
            &text[lo..text.len()],
            lo,
            text.len(),
        ));
    }

    slices
}

/// Produce an `SGR` from a styling sequence.
///
/// Requires `Match` as we can assume skipping of certain bytes.
fn handle_seq(m: &Match) -> SGR {
    // the slice we want to process is skipped of first two bytes (ESC[) and last byte (terminating byte)
    let slice = &m.text[2..(m.text.len() - 1)];

    let styles = slice.split(SEPARATOR);

    let mut sgr = SGR::default();

    for style in styles {
        adjust_sgr(&mut sgr, style);
    }

    sgr
}

/// Apply the style seq to the SGR. Maps decimal numbers according to
/// spec at https://en.wikipedia.org/wiki/ANSI_escape_code#Escape_sequences.
fn adjust_sgr(sgr: &mut SGR, seq: &str) {
    match seq {
        "0" => *sgr = SGR::default(),                  // 0
        "1" => sgr.intensity = Intensity::Bold,        // 1
        "2" => sgr.intensity = Intensity::Faint,       // 2
        "3" => sgr.italic = true,                      // 3
        "4" => sgr.underline = true,                   // 4
        "5" => sgr.blink = true,                       // 5
        "7" => sgr.reversed = true,                    // 7
        "8" => sgr.hidden = true,                      // 8
        "9" => sgr.strikethrough = true,               // 9
        "22" => sgr.intensity = Intensity::Normal,     // 22
        "23" => sgr.italic = false,                    // 23
        "24" => sgr.underline = false,                 // 24
        "25" => sgr.blink = false,                     // 25
        "27" => sgr.reversed = false,                  // 27
        "28" => sgr.hidden = false,                    // 28
        "29" => sgr.strikethrough = false,             // 29
        "30" => sgr.fg_colour = Color::Black,          // 30
        "31" => sgr.fg_colour = Color::Red,            // 31
        "32" => sgr.fg_colour = Color::Green,          // 32
        "33" => sgr.fg_colour = Color::Yellow,         // 33
        "34" => sgr.fg_colour = Color::Blue,           // 34
        "35" => sgr.fg_colour = Color::Magenta,        // 35
        "36" => sgr.fg_colour = Color::Cyan,           // 36
        "37" => sgr.fg_colour = Color::White,          // 37
        "40" => sgr.bg_colour = Color::Black,          // 40
        "41" => sgr.bg_colour = Color::Red,            // 41
        "42" => sgr.bg_colour = Color::Green,          // 42
        "43" => sgr.bg_colour = Color::Yellow,         // 43
        "44" => sgr.bg_colour = Color::Blue,           // 44
        "45" => sgr.bg_colour = Color::Magenta,        // 45
        "46" => sgr.bg_colour = Color::Cyan,           // 46
        "47" => sgr.bg_colour = Color::White,          // 47
        "90" => sgr.fg_colour = Color::BrightBlack,    // 90
        "91" => sgr.fg_colour = Color::BrightRed,      // 91
        "92" => sgr.fg_colour = Color::BrightGreen,    // 92
        "93" => sgr.fg_colour = Color::BrightYellow,   // 93
        "94" => sgr.fg_colour = Color::BrightBlue,     // 94
        "95" => sgr.fg_colour = Color::BrightMagenta,  // 95
        "96" => sgr.fg_colour = Color::BrightCyan,     // 96
        "97" => sgr.fg_colour = Color::BrightWhite,    // 97
        "100" => sgr.bg_colour = Color::BrightBlack,   // 100
        "101" => sgr.bg_colour = Color::BrightRed,     // 101
        "102" => sgr.bg_colour = Color::BrightGreen,   // 102
        "103" => sgr.bg_colour = Color::BrightYellow,  // 103
        "104" => sgr.bg_colour = Color::BrightBlue,    // 104
        "105" => sgr.bg_colour = Color::BrightMagenta, // 105
        "106" => sgr.bg_colour = Color::BrightCyan,    // 106
        "107" => sgr.bg_colour = Color::BrightWhite,   // 107
        _ => (),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use colored::Colorize;

    #[test]
    fn cat_and_matches_len() {
        let txt = "hello";
        let matches = parse(&txt);
        let cat = categorise_text(&txt);
        assert!(matches.len() + 1 >= cat.len());

        let txt = "hello".bright_green();
        let matches = parse(&txt);
        let cat = categorise_text(&txt);
        assert!(matches.len() + 1 >= cat.len());

        let txt = format!("{}{}{}", "hello".bright_green(), "world".red(), "whatever");
        let matches = parse(&txt);
        let cat = categorise_text(&txt);
        assert!(matches.len() + 1 >= cat.len());
    }
}
